### PR TITLE
8289822: G1: Make concurrent mark code owner of TAMSes

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  please review this change that moves TAMS ownership and handling out of `HeapRegion` into `G1ConcurrentMark`.

The rationale is that it's intrinsically a data structure only used by the marking, and only interesting during marking, so attaching it to the always present `HeapRegion` isn't fitting.

This also moves some code that is about marking decisions out of `HeapRegion`.

In this change the TAMSes are still maintained and kept up to date as long as the `HeapRegion` exists (basically always) as the snapshotting of the marking does not really snapshot ("copy over the `HeapRegion`s that existed at the time of the snapshot, only ever touching them), but assumes that they exist even for uncommitted (or free) regions due to the way it advances the global `finger` pointer.
I did not want to change this here.

This is also why `HeapRegion` still tells `G1ConcurrentMark` to update the TAMS in two places (when clearing a region and when resetting during full gc).

Another option I considered when implementing this is clearing all marking statistics (`G1ConcurrentMark::clear_statistics`) at these places in `HeapRegion` because it decreases the amount of places where this needs to be done (maybe one or the other place isn't really necessary already). I rejected it because this would add some work that is dependent on the number of worker threads (particularly) into `HeapRegion::hr_clear()`. This change can be looked at by viewing at all but the last commit.

Testing: tier1-4, gha

Thanks,
  Thomas

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289822](https://bugs.openjdk.org/browse/JDK-8289822): G1: Make concurrent mark code owner of TAMSes (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [fae12ca0](https://git.openjdk.org/jdk/pull/18150/files/fae12ca00f71eed47c3a98413885cf79678a1abd)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**) ⚠️ Review applies to [fae12ca0](https://git.openjdk.org/jdk/pull/18150/files/fae12ca00f71eed47c3a98413885cf79678a1abd)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18150/head:pull/18150` \
`$ git checkout pull/18150`

Update a local copy of the PR: \
`$ git checkout pull/18150` \
`$ git pull https://git.openjdk.org/jdk.git pull/18150/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18150`

View PR using the GUI difftool: \
`$ git pr show -t 18150`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18150.diff">https://git.openjdk.org/jdk/pull/18150.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18150#issuecomment-1983300127)

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1ConcurrentMark.cpp [hunk 1]
```diff
@@ -584,7 +589,7 @@ void G1ConcurrentMark::humongous_object_eagerly_reclaimed(HeapRegion* r) {
 
   if (!_g1h->collector_state()->mark_or_rebuild_in_progress()) {
     return;
-  }
+}
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1ConcurrentMark.cpp:592]: Seem unnecessary.

File: src/hotspot/share/gc/g1/g1ConcurrentMark.hpp [hunk 2]
```diff
@@ -537,6 +538,9 @@ class G1ConcurrentMark : public CHeapObj<mtGC> {
 
   // Region statistics gathered during marking.
   G1RegionMarkStats* _region_mark_stats;
+  // Top pointer for each region at the start of marking. Must be valid for all committed
+  // regions.
+  HeapWord* volatile* _top_at_mark_starts;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1ConcurrentMark.hpp:543]: I wonder if one can use `HeapWord* volatile _top_at_mark_starts[];` to emphasize that it's an array.

File: src/hotspot/share/gc/g1/g1ConcurrentMark.hpp [hunk 3]
```diff
@@ -556,6 +560,16 @@ class G1ConcurrentMark : public CHeapObj<mtGC> {
   // Set live bytes for concurrent marking.
   void set_live_bytes(uint region, size_t live_bytes) { _region_mark_stats[region]._live_words = live_bytes / HeapWordSize; }
 
+  // Update the TAMS for the given region to the current top.
+  inline void update_top_at_mark_start(HeapRegion* r);
+  // Reset the TAMS for the given region to bottom of that region.
+  inline void reset_top_at_mark_start(HeapRegion* r);
+
+  inline HeapWord* top_at_mark_start(const HeapRegion* r) const;
+  inline HeapWord* top_at_mark_start(uint region) const;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1ConcurrentMark.hpp:569]: Maybe call it `region_index` or sth alike?  (In the impl of these methods, having `region` being the index is a bit confusing, IMO.)

File: src/hotspot/share/memory/iterator.hpp [hunk 4]
```diff
@@ -41,7 +41,7 @@ class Thread;
 
 // The following classes are C++ `closures` for iterating over objects, roots and spaces
 
-class Closure : public StackObj { };
+class Closure : public CHeapObj<mtGC> { };
```
Comment (albertnetymk) [inline src/hotspot/share/memory/iterator.hpp:44]: Why is this required (in this PR)?

**Review by albertnetymk — COMMENTED** [review]
Maybe add `G1: ` in the title?

File: src/hotspot/share/memory/iterator.hpp (hunk 4 above)
Comment (tschatzl) [reply src/hotspot/share/memory/iterator.hpp:44]: The changes need to allocate `G1CollectedHeap::_is_alive_closure_cm` when creating that argument for creating the concurrent reference processor. Previously the closure used `G1CollectedHeap` which is available when inlining it into the `G1CollectedHeap` instance, but now the closure needs `G1ConcurrentMark` which is not available at that time (i.e. null).
The alternative would have been introducing something like an `initialize` method to that closure which I did not like.

File: src/hotspot/share/gc/g1/g1ConcurrentMark.hpp (hunk 2 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1ConcurrentMark.hpp:543]: No, it would be interpreted as flexible array, which would need to be the last member of the class.

File: src/hotspot/share/memory/iterator.hpp (hunk 4 above)
Comment (albertnetymk) [reply src/hotspot/share/memory/iterator.hpp:44]: OK, could `G1CMIsAliveClosure::do_object_b` retrieve `_cm` via `_g1h`? (It's obvious why `Closure` is part of mtGC here.)

Comment (tschatzl) [reply src/hotspot/share/memory/iterator.hpp:44]: At the cost of an additional indirection per reference. I would prefer to just have an `initialize` method then.
You are right that just using `mtGC` is not appropriate here as it is used in non-GC code too. Is an `initialize` method okay with you then?

File: src/hotspot/share/memory/iterator.hpp [hunk 5]
```diff
@@ -41,7 +41,7 @@ class Thread;
 
 // The following classes are C++ `closures` for iterating over objects, roots and spaces
 
-class Closure : public StackObj { };
+class Closure { };
```
Comment (albertnetymk) [inline src/hotspot/share/memory/iterator.hpp:44]: Why is this required?

Comment (tschatzl) [reply src/hotspot/share/memory/iterator.hpp:44]: I just forgot to undo to the original code...

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp [hunk 6]
```diff
@@ -1506,6 +1506,7 @@ void G1CollectedHeap::ref_processing_init() {
   //     * Reference discovery will not need a barrier.
 
   // Concurrent Mark ref processor
```
Comment (walulyai) [inline src/hotspot/share/gc/g1/g1CollectedHeap.cpp:1508]: The comment now seems misplaced

**Review by walulyai — APPROVED** [review]
LGTM!

File: src/hotspot/share/gc/g1/g1CollectedHeap.cpp (hunk 6 above)
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CollectedHeap.cpp]: Moved back the comment to the original position.

**Comment by tschatzl** [comment]
Thanks @albertnetymk @walulyai for your reviews
/integrate
